python/ray/ml/preprocessor.py

```diff
@@ -1,55 +1,101 @@
+        """Fit this Preprocessor to the Dataset.
 
-        Args:
-            ds (Dataset): Input dataset.
+        Fitted state attributes will be directly set in the Preprocessor.
 
         Returns:
-            Dataset: Transformed dataset.
-
+            Preprocessor: The fitted Preprocessor with state attributes.
         """
-        raise NotImplementedError
+        if self.check_is_fitted():
+            raise PreprocessorAlreadyFittedException(
+                "`fit` cannot be called multiple times. "
+                "Create a new Preprocessor to fit a new Dataset."
+            )
 
-    def transform(self, ds: Dataset) -> Dataset:
-        """Transforms the given dataset.
+        return self._fit(dataset)
 
-        The preprocessor must be fitted first.
+    def _fit(self, dataset: Dataset) -> "Preprocessor":
+        raise NotImplementedError()
 
-        Args:
-            ds (Dataset): Input dataset.
+    def fit_transform(self, dataset: Dataset) -> Dataset:
+        """Fit this Preprocessor to the Dataset and then transform the Dataset.
 
+        Args:
+            dataset (Dataset): Input Dataset.
         Returns:
-            Dataset: Transformed dataset.
+            Dataset: The transformed Dataset.
+        """
+        self.fit(dataset)
+        return self.transform(dataset)
 
+    def transform(self, dataset: Dataset) -> Dataset:
+        """Transform the given dataset.
+
+        Args:
+            dataset (Dataset): Input Dataset.
+        Returns:
+            Dataset: The transformed Dataset.
         """
-        raise NotImplementedError
+        if self._is_fittable and not self.check_is_fitted():
+            raise PreprocessorNotFittedException(
+                "`fit` must be called before `transform`."
+            )
+        return self._transform(dataset)
+
+    def _transform(self, dataset: Dataset) -> Dataset:
+        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
 
     def transform_batch(self, df: DataBatchType) -> DataBatchType:
         """Transform a single batch of data.
 
-        The preprocessor must be fitted first.
-
         Args:
             df (DataBatchType): Input data batch.
-
         Returns:
-            Dataset: Transformed data batch.
+            DataBatchType: The transformed data batch.
         """
-        raise NotImplementedError
 
-    def __getstate__(self) -> dict:
-        return self.__dict__
+        if self._is_fittable and not self.check_is_fitted():
+            raise PreprocessorNotFittedException(
+                "`fit` must be called before `transform_batch`."
+            )
+        return self._transform_batch(df)
+
+    def _transform_batch(self, df: DataBatchType) -> DataBatchType:
+        import pandas as pd
+
+        if not isinstance(df, pd.DataFrame):
+            raise NotImplementedError(
+                "`transform_batch` is currently only implemented for Pandas DataFrames."
```

Comment:
Eventually, it might be nice for the preprocessors to be able to provide multiple batch transformation implementations (e.g. both Arrow and Pandas) which would allow the preprocessor to dynamically choose either implementation based on the input dataset format. This way, if the dataset blocks are already in Pandas DataFrame form, as an example, we can choose the Pandas DataFrame batch transformation implementation, which will allow us to avoid the copy/conversion cost between Arrow and Pandas.

Reply:
Makes sense, adding a TODO for this!